cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/PageRank.scala

```diff
@@ -44,15 +44,28 @@ case class PageRankIterationState(pageRank: Array[Double], error: Double, iterat
 class PageRank(graph: DirectedGraph[Node], params: PageRankParams = PageRankParams())
   extends LinkAnalysis[PageRankIterationState](graph, params, "pagerank") {
 
-  private val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private lazy val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private val nodeCount = graph.nodeCount
 
-  graph foreach { node => efficientNeighbors(node) foreach { nbr => outboundCount(nbr) += 1} }
+  private val dangling = if (isInStored) {
+    val nonDangling = graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      val neighbors = efficientNeighbors(node)
+      neighbors foreach { nbr => outboundCount(nbr) += 1 }
+      partialSet ++ neighbors.toSet
+    }
+    graph.map { _.id }.toSet -- nonDangling
+  }
+  else {
+    graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      if (efficientNeighbors(node).isEmpty) partialSet + node.id else partialSet
+    }
+  }
```

Comment:
I have a question--this could be a design choice or a bug and want to know which one it is.  Lets take `TestGraphs.dangling_g7_in` as an example which is a graph with man zeroOutgoing nodes and only stores the in bound direction.  NodeId 2 has two inbound connections (3, 4) and itself forms an inbound connection to (1, 6).  In my mind, the following code should pass a test:

```scala
   val graph = TestGraphs.dangling_g7_in
   val node2 = graph.getNodeById(2).get
   node2.inboundCount shouldEqual 2
   node2.outboundCount shouldEqual 2
```
However, the code in its current state returns `node2.outboundCount` as being 0 since it does not store that connection.

Was this chosen to prevent an expensive computation from occuring?  Obviously to get the outbound connections for a graph stored OnlyIn would be much more costly because one could potentially parse every node looking for incident connection.  If you could clear this up for me that would be much appreciated.